cmd/main.go

```diff
@@ -270,6 +279,71 @@ func run(opts *options) error {
 		return err
 	}
 
+	// --failed-first logic
+	if opts.failedFirstJSONFile != "" {
+		data, err := os.ReadFile(opts.failedFirstJSONFile)
+		if err != nil {
+			return fmt.Errorf("failed to read previous run JSON file: %w", err)
+		}
+		// Parse JSON lines
+		var failedTests []struct{ Package, Test, Action string }
+		for _, line := range strings.Split(string(data), "\n") {
+			if strings.TrimSpace(line) == "" {
+				continue
+			}
+			var evt struct{ Package, Test, Action string }
+			_ = json.Unmarshal([]byte(line), &evt)
+			if evt.Action == "fail" && evt.Test != "" {
+				failedTests = append(failedTests, evt)
+			}
+		}
```

Comment:
There's code in the `testjson` package to do this: https://pkg.go.dev/gotest.tools/gotestsum@v1.12.2/testjson#ScanTestOutput

I'd suggest using that to read the JSON.